Add Zod Validation to Backend API Routes and fix notes e2e tests

tsahil01 opened the pull request:
#411

### **What Changed**
Added Zod validation schemas to multiple API routes for better input validation and type safety.

### **Why This Change Was Needed**
**Before:** 
- Manual validation with basic JavaScript checks (`typeof`, `!value`, string trimming)
- Repetitive validation code in every API route
- No type safety for request bodies
- Generic error messages that didn't help developers
- Error in notes e2e tests wrt to `NOTE_COLORS`

**After:**
- Strong type-safe validation using Zod schemas
- Consistent validation patterns across all routes
- Clear, specific error messages for invalid inputs
- Better developer experience with autocomplete and type checking
- Notes e2e tests properly follows `NOTE_COLORS`

slavingia commented on app/api/boards/[id]/notes/[noteId]/route.ts:
```diff
@@ -1,14 +1,31 @@
 } from "@/lib/slack";
 
-// Update a note
+const checklistItemSchema = z.object({
```
putting types in one file may be nice.

slavingia commented on app/api/boards/all-notes/notes/route.ts:
```diff
@@ -1,8 +1,20 @@
+  checked: z.boolean().optional().default(false),
+  order: z.number().optional(),
+});
+const createGlobalNoteSchema = z.object({
```
I would just same these noteSchema etc. Should not be specific to output, input, etc generally

tsahil01 commented:
@slavingia Done with the requested changes.

slavingia commented on lib/types/zod-types.ts:
```diff
@@ -0,0 +1,45 @@
+import { z } from "zod";
```
Call this index.ts?

ershad commented on lib/types/index.ts:
```diff
@@ -0,0 +1,45 @@
+
+export const noteSchema = z.object({
+  boardId: z.string(),
+  color: z.string().optional(),
```
Would be good to validate against `NOTE_COLORS`?

ershad commented on app/api/boards/[id]/notes/[noteId]/route.ts:
```diff
@@ -69,36 +89,15 @@ export async function PUT(
       sanitizedChecklistItems = [...checklistItems]
         .sort((a, b) => a.order - b.order)
-        .map((item, i) => ({ ...item, order: i }));
+        .map((item, i) => ({ ...item, order: i, id: item.id || "" }));
```
Shouldn't ideally set `id` as `""`?

ershad commented on app/api/boards/route.ts:
```diff
@@ -86,15 +88,32 @@ export async function POST(request: NextRequest) {
+        .extend({
+          name: z
+            .string()
+            .min(1, "Board name is required and cannot be empty or only whitespace")
+            .refine(
+              (val) => val.trim().length > 0,
+              "Board name is required and cannot be empty or only whitespace"
```
duplicate validations?

ershad commented on app/api/boards/[id]/route.ts:
```diff
@@ -118,9 +139,8 @@ export async function PUT(request: NextRequest, { params }: { params: Promise<{
-    if (description !== undefined)
-      updateData.description = description?.trim() || board.description;
+    if (name !== undefined) updateData.name = name.trim() || board.name;
+    if (description !== undefined) updateData.description = (description?.trim() || null) as string;
```
would be nice to remove the `null as string` assertion.

ershad reviewed (changes requested):
@tsahil01 _a thanks for the PR! Added a few comments, please take a look.

tsahil01 commented on lib/types/index.ts:
```diff
@@ -0,0 +1,46 @@
+
+export const noteSchema = z.object({
+  boardId: z.string(),
+  color: z.enum(NOTE_COLORS).optional(),
```
@ershad 
To be able to work this, I fixed the `notes.spec.ts` colors as they were different from `NOTE_COLORS`  
this was causing test inconsistencies with the actual color validation schema. All hardcoded colors now match the centralized constants.

ershad replied:
awesome, thanks!

tsahil01 commented:
@ershad Done.
made the suggested changes. 
now checklist item id required in zod schema, 
removed fallback uuid generation in API route, 
fixed copy functionality in `app/boards/[id]/page.tsx` to generate IDs for checklist items.

ershad commented on app/api/boards/[id]/notes/[noteId]/route.ts:
```diff
@@ -69,36 +89,15 @@ export async function PUT(
       sanitizedChecklistItems = [...checklistItems]
         .sort((a, b) => a.order - b.order)
-        .map((item, i) => ({ ...item, order: i }));
+        .map((item, i) => ({ ...item, order: i, id: item.id }));
```
```suggestion
        .map((item, i) => ({ ...item, order: i }));
```

ershad reviewed (approved):
@tsahil01 thanks for the PR!

ershad commented:
@tsahil01 please reach out to @slavingia regarding the bounty.
